Derive Eq for std::cmp::Ordering, instead of using manual impl.

zachs18 opened the pull request:
This allows consts of type Ordering to be used in patterns, and with feature(adt_const_params) allows using `Ordering` as a const generic parameter.

Currently, `std::cmp::Ordering` implements `Eq` using a manually written `impl Eq for Ordering {}`, instead of `derive(Eq)`. This means that it does not implement `StructuralEq`.

This commit removes the manually written impl, and adds `derive(Eq)` to `Ordering`, so that it will implement `StructuralEq`.

zachs18 committed:
Derive Eq for std::cmp::Ordering, instead of using manual impl.

This allows consts of type Ordering to be used in patterns, and (with feature(adt_const_params)) allows using Orderings as const generic parameters.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @kennytm (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

zachs18 committed:
Add test for StructuralEq for std::cmp::Ordering.

Added test in library/core/tests/cmp.rs that ensures that `const`s of type `Ordering`s can be used in patterns.

zachs18 commented:
The manual impl appears to have been added in 2013 with this commit ( https://github.com/rust-lang/rust/commit/a0f8540c9571dadb9af542799156cc8a96263c41 ). As far as I can tell, `derive(Eq)` (`deriving(TotalEq)`) was possible then, so I don't know if there is a reason it was not derived at the time.

Dylan-DPC commented:
r? @Dylan-DPC

Dylan-DPC commented:
@bors r+ rollup

m-ou-se commented:
This needed a @rust-lang/libs-api FCP, as this was a change to the public stable API.

This change has already shipped stably as part of 1.61. Hopefully everyone agrees this change was uncontroversial.

m-ou-se commented:
This adds an implementation of `StructuralEq`. (See the top comment.)

BurntSushi commented:
Oh wow, okay, that's subtle. Hopefully we can figure out how to make such changes more explicit in the future.

But I _think_ this specific change looks okay at least.
